File: ones-gfx-sdk-python/ones_gfx/resources/fabrics.py
from __future__ import annotations

from typing import Any

from ..models import Fabric
from ..transport import Transport


class FabricsResource:
    def __init__(self, transport: Transport):
        self._transport = transport

    def list(self) -> list[Fabric]:
        body = self._transport.get("fabrics")
        items = body.get("fabrics", []) if isinstance(body, dict) else []
        return [Fabric.from_api(item) for item in items]

    def modify_gpu_allocations(
        self,
        fabric_name: str,
        tenant_name: str,
        operation: str,
        suid: dict[str, dict[str, dict[str, list[str]]]],
    ) -> dict[str, Any]:
        """Map or unmap specific GPUs to a tenant using fine-grained suid addressing.

        POST /fabrics/{fabricName}/tenants/{tenantName}/gpuAllocations

        Args:
            fabric_name:  Target fabric.
            tenant_name:  Target tenant.
            operation:    ``"ADD"`` to map GPUs or ``"DELETE"`` to unmap.
            suid:         Nested map: server_index -> hostname -> {"gpus": ["G0", ...]}.

        Example::

            client.fabrics.modify_gpu_allocations(
                fabric_name="fab1",
                tenant_name="tenant1",
                operation="ADD",
                suid={
                    "0": {
                        "hgx-su00-h00": {"gpus": ["G0", "G1", "G2", "G3"]}
                    }
                },
            )

        Returns:
            dict with keys ``status``, ``operationId`` (if async), ``message``.
        """
        if not fabric_name:
            raise ValueError("fabric_name is required")
        if not tenant_name:
            raise ValueError("tenant_name is required")
        if operation not in ("ADD", "DELETE"):
            raise ValueError(f'operation must be "ADD" or "DELETE", got {operation!r}')
        if not suid:
            raise ValueError("suid must contain at least one server entry")

        body = {"operation": operation, "suid": suid}
        path = f"fabrics/{fabric_name}/tenants/{tenant_name}/gpuAllocations"
        result = self._transport.post(path, json_body=body)
        if isinstance(result, dict):
            return result
        if isinstance(result, str):
            return {"status": "success", "message": result}
        if result is None:
            return {"status": "success"}
        raise TypeError(
            f"unexpected response type for modify_gpu_allocations: {type(result).__name__}"
        )

File: ones-gfx-sdk-go/ones_gfx/resources/fabrics.go
package resources

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/aviznetworks/ones-gfx-sdk/ones-gfx-sdk-go/ones_gfx"
)

// FabricsResource handles operations on the fabrics collection.
type FabricsResource struct {
	transport *ones_gfx.Transport
}

// NewFabricsResource constructs a FabricsResource.
func NewFabricsResource(transport *ones_gfx.Transport) *FabricsResource {
	return &FabricsResource{transport: transport}
}

// List retrieves all fabrics defined in the ONES system.
// Maps to GET /fabrics.
func (r *FabricsResource) List(ctx context.Context) ([]ones_gfx.Fabric, error) {
	body, err := r.transport.Get("fabrics", nil)
	if err != nil {
		return nil, err
	}

	// Response shape: {"fabrics": [ {...}, {...} ]}
	obj, ok := body.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", body)
	}

	fabricsRaw, ok := obj["fabrics"]
	if !ok {
		return nil, fmt.Errorf("response missing 'fabrics' field")
	}

	fabricsList, ok := fabricsRaw.([]interface{})
	if !ok {
		return nil, fmt.Errorf("'fabrics' field is not an array")
	}

	fabrics := make([]ones_gfx.Fabric, 0, len(fabricsList))
	for _, item := range fabricsList {
		jsonBytes, err := remarshal(item)
		if err != nil {
			continue
		}
		var fabric ones_gfx.Fabric
		if err := unmarshalJSON(jsonBytes, &fabric); err != nil {
			continue
		}
		fabrics = append(fabrics, fabric)
	}

	return fabrics, nil
}

// ModifyGPUAllocations maps or unmaps specific GPUs to a tenant on shared
// fabric servers. It maps to:
//
//	POST /fabrics/{fabricName}/tenants/{tenantName}/gpuAllocations
//
// Use req.Operation = ones_gfx.OperationAdd to map GPUs, or
// ones_gfx.OperationDelete to remove the mapping. The Suid field targets one
// or more servers by index and hostname, each with an explicit list of GPU IDs.
//
// Example:
//
//	resp, err := client.Fabrics.ModifyGPUAllocations(ctx, "fabric1", "tenant1",
//	    ones_gfx.GPUAllocationRequest{
//	        Operation: ones_gfx.OperationAdd,
//	        Suid: map[string]map[string]ones_gfx.ServerGPUs{
//	            "0": {"hgx-su00-h00": {GPUs: []string{"G0", "G1", "G2", "G3"}}},
//	        },
//	    })
func (r *FabricsResource) ModifyGPUAllocations(
	ctx context.Context,
	fabricName string,
	tenantName string,
	req ones_gfx.GPUAllocationRequest,
) (*ones_gfx.GPUAllocationResponse, error) {
	if fabricName == "" {
		return nil, fmt.Errorf("fabricName is required")
	}
	if tenantName == "" {
		return nil, fmt.Errorf("tenantName is required")
	}
	if err := req.Validate(); err != nil {
		return nil, err
	}

	path := fmt.Sprintf("fabrics/%s/tenants/%s/gpuAllocations", fabricName, tenantName)
	result, err := r.transport.Post(path, req, ones_gfx.OperationModeSynchronous, nil)
	if err != nil {
		return nil, err
	}

	if msg, ok := result.(string); ok {
		return &ones_gfx.GPUAllocationResponse{
			Status:  "success",
			Message: msg,
		}, nil
	}

	jsonBytes, err := json.Marshal(result)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal GPU allocation response: %w", err)
	}

	var resp ones_gfx.GPUAllocationResponse
	if err := json.Unmarshal(jsonBytes, &resp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal GPU allocation response: %w", err)
	}

	return &resp, nil
}
